### PR TITLE
Rework to support cancelling fetch requests

### PR DESCRIPTION
**Pull request by jonkoops**

### CODE REVIEW

File: src/usePromise.ts [hunk 1]
```diff
@@ -1,75 +1,89 @@
...
-export interface PromiseRejectedResult {
-  status: PromiseStatus.Rejected
-  error: any
-}
+/**
+ * Determines if the result of a promise is pending.
+ * @param result Result to check.
+ */
+export const isPending = (
+  result: PromiseResult<any>,
```
Comment (samuelleeuwenburg) [inline src/usePromise.ts:19]: can be changed to:

```
export const isPending = <T>(
  result: PromiseResult<T>,
```

File: src/usePromise.ts [hunk 2]
```diff
@@ -1,75 +1,89 @@
...
+): result is PromisePendingResult => result.status === 'pending'
 
-export type PromiseFactoryFn<T> = () => Promise<T>
+/**
+ * Determines if the result of a promise is fulfilled.
+ * @param result Result to check.
+ */
+export const isFulfilled = <T = any>(
```
Comment (samuelleeuwenburg) [inline src/usePromise.ts:26]: can be changed to `<T>`

File: src/usePromise.ts [hunk 3]
```diff
@@ -1,75 +1,89 @@
...
+): result is PromiseFulfilledResult<T> => result.status === 'fulfilled'
 
 /**
- * Takes a function that creates a promise and returns its resolved or rejected value together with the status of the promise.
- * When passing in a promise make sure that the dependencies are passed as well:
+ * Determines if the result of a promise is rejected.
+ * @param result Result to check.
+ */
+export const isRejected = (
+  result: PromiseResult<any>,
```
Comment (samuelleeuwenburg) [inline src/usePromise.ts:35]:
```
export const isRejected = <T>(
  result: PromiseResult<T>,
 ```

File: src/usePromise.ts [hunk 4]
```diff
@@ -1,75 +1,89 @@
...
+ * Can abort a fetch request, a [signal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) is provided from the factory function to do so.
+ *
+ * ```ts
+ * const result = usePromise(signal => fetch(`/api/products/${id}`, { signal }), [id])
+ * ```
+ *
+ * @param factory Function that creates the promise.
+ * @param deps If present, promise will be recreated if the values in the list change.
  */
 export default function usePromise<T = any>(
```
Comment (samuelleeuwenburg) [inline src/usePromise.ts:60]: Same here, `<T>`

File: README.md [hunk 5]
```diff
@@ -16,84 +22,11 @@ Or [`npm`](https://www.npmjs.com/package/@amsterdam/use-promise):
...
-
-First off, we pass a function as the first argument. This is the factory function that is responsible for creating our promise.
-
-This function is called when our component is mounted, or if the dependencies passed in as the second argument change. This is essentially the same as the [useMemo](https://reactjs.org/docs/hooks-reference.html#usememo) hook.
-
-The `usePromise` function returns a special object that contains the `status` of the Promise and some other fields. Using this `status` field we can determine what to show the user.
+## Guides
+- [Getting Started](./guides/getting-started.md)
+- [Handling Side-effects](./guides/side-effects.md)
+- [Cancelling Fetch Requests](./guides/cancelling-requests.md)
```
Comment (samuelleeuwenburg) [inline README.md:28]: Cool way to split docs!

**Comment by samuelleeuwenburg** [comment]
Also wondering if it is perhaps possible to put the status in an enum type to help cover all the possibilities

**Comment by jonkoops** [comment]
> Also wondering if it is perhaps possible to put the status in an enum type to help cover all the possibilities

Yeah, this is what I thought originally. Unfortunately this will clash with the built-in `PromiseSettledResult` as it uses a string type for the `status` field. Even if the enum is string indexed it cannot be directly assigned anymore, see https://github.com/microsoft/TypeScript/issues/17690

File: README.md (hunk 5 above)
Comment (jonkoops) [reply README.md:28]: Thanks!

File: README.md [hunk 6]
```diff
@@ -16,84 +22,11 @@ Or [`npm`](https://www.npmjs.com/package/@amsterdam/use-promise):
...
-
-On the last line of our component we can now safely assume that the status is `Resolved` as we already checked all other possible statuses. This means that we can now also access the `value` property to get the resolved value of the Promise.
-
-```tsx
-return <p>You're looking at the user information of ${result.value.firstName}!<p>
-```
+## Compatibility
 
-That's it, you're now the master of Promises! 💪
+This library will work in every environment (Node or browser) that supports [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController), [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal), [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) and [async functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function). If you need to support an environment that does not have these features make sure to include the appropriate polyfills and transformations.
```
Comment (chrisvanmook) [inline README.md:32]: Perhaps refer to babel for transformations / polyfills? As this is one of the most used compiler

File: src/usePromise.ts [hunk 7]
```diff
@@ -1,75 +1,89 @@
...
-}
-
-export type PromiseResult<T> =
-  | PromisePendingResult
-  | PromiseFulfilledResult<T>
-  | PromiseRejectedResult
+export type PromiseResult<T> = PromisePendingResult | PromiseSettledResult<T>
 
 export interface PromisePendingResult {
-  status: PromiseStatus.Pending
```
Comment (chrisvanmook) [inline src/usePromise.ts:15]: Why aren't we using enums anymore?

**Review by chrisvanmook — APPROVED** [review]

File: src/usePromise.ts (hunk 7 above)
Comment (jonkoops) [reply src/usePromise.ts:15]: Because of: https://github.com/Amsterdam/use-promise/pull/6#issuecomment-808939120

File: README.md (hunk 6 above)
Comment (jonkoops) [reply README.md:32]: I'd like to keep this out for now as there should be plenty of resources around for users on how to do this. Most tooling such a Create React App and Next will do this for you automatically.

**Pull request merged**